### DISCUSSION TITLE
zh/editor

### DISCUSSION THREAD

**Discussion by giscus** [casbin/casbin#1220] [Docs comments]
# zh/editor

Casbin Online Editor

https://casbin.org/zh/editor

**Comment by abhijithp** [comment]
there seems to be an error with the result 
for request: 
alice, read, data1
alice, write, data1
bob, write, data2
bob, read, data2
bob, write, data1

with policy
p, alice, reader, data1
p, bob, owner, data2

g, reader, read
g, owner, read
g, owner, write

**Reply by JohnHuang-2023** [reply]
I use the following RBAC model and the results are true, false, true, true, false.
```
[request_definition]
r = sub, act, obj

[policy_definition]
p = sub, act, obj

[role_definition]
g = _, _
g2 = _, _

[policy_effect]
e = some(where (p.eft == allow))

[matchers]
m = r.sub == p.sub && g(p.act, r.act) && r.obj == p.obj
```

**Reply by phamkhai11123** [reply]
[request_definition]
r = sub, obj, act
[policy_definition]
p = sub, obj, act
[role_definition]
g = _,_ 
[policy_effect]
e = some(where (p.eft == allow))
[matchers]
m = g(r.sub, p.sub) && r.obj == p.obj && (r.act == p.act || p.act == "*")
this config won't make you disaponitend

**Comment by abhijithp** [comment]
the result is getting as all false

**Comment by wuchencm** [comment]
https://editor.casbin.org/#H6WBBZZ3J
How should I implement the judgment when a user belongs to multiple groups?

**Comment by zaytsevand** [comment]
As of today, the Editor doesn't seem to work: Enforcement Result never changes.

**Reply by gutakk** [reply]
I also noticed this. I can't find any button to make a request

**Comment by Tbag12** [comment]
```
[request_definition]
r = sub, obj, act

[policy_definition]
p = sub, obj, act

[role_definition]
g = _, _

[policy_effect]
e = some(where (p.eft == allow))

[matchers]
m = g(r.sub, p.sub) && r.obj == p.obj && r.act == p.act
```
---
```
g, admin, *
p, admin, *, *
p, bob, /api/v1/a, GET
p, bob, /api/v1/a, GET
```

How to use wildcards to judge? I want to use * to indicate that admin has all permissions, but `Enforce(sub,ojb,act)` always returns false

**Reply by Tbag12** [reply]
who can help me?

**Reply by hsluoyz** [reply]
@Tbag12 go to editor: https://editor.casbin.org/, click "Ask AI"

**Reply by inquisitorOzzak** [reply]
You want to put the wildcards in the regex like so:

[matchers]
m = g(r.sub, p.sub) && (r.obj == p.obj || p.obj =="*") && (r.act == p.act || p.act == "*")

**Comment by anhtuank7c** [comment]
Thank you so much for creating a very helpful library and online editor to design and test.

I struggled at first but now I seem to understand it since mock policies look hard to understand to me.

So I created 2 examples to learn casbin.

* RBAC: https://gist.github.com/anhtuank7c/eeec11182ee2290b9de285b9456b624c
* RBAC with Resource roles: https://gist.github.com/anhtuank7c/56ce024c060fb5d34cbb921f5ddd4da3

**Comment by vishesh-bimacred** [comment]
What is this "g" in the role_definition and where is this function defined  m = g(r.sub, p.sub)?? 

It takes two arguments but what does it do?

**Reply by anhtuank7c** [reply]
https://gist.github.com/anhtuank7c/eeec11182ee2290b9de285b9456b624c#file-rbac-L34
Let consider `g` is to define a group of user
IE:
```
g, username_01, group_admin
g, username_02, group_admin

g, username_01, group_accountant
g, username_03, group_accountant
```

So we have 2 group: `group_admin` and `group_accountant`

And I assigned `username_01` to **both group**, `username_02` to **group_admin**, `username_03` to **group_accountant**

I hope this can help.

**Comment by z33gg** [comment]
Node-Casbin engine and Casbin-rs engine have different result in this case。
model.conf
```
[request_definition]
r = sub, obj

[policy_definition]
p = sub, obj, eft

[role_definition]
g = _, _
g2 = _,_

[policy_effect]
e = some(where (p.eft == allow))

[matchers]
m =  (g(r.sub, p.sub) && r.obj == p.obj ) && (g2(r.sub, p.sub) && r.obj == p.obj )


```
policy.csv
```
p, organization_alice, data1, allow
p, organization_alice_1, data1, deny
p, role_alice, data1, allow
p, role_alice, data2, allow
g, alice, role_alice
g2,alice, organization_alice_1
g2,organization_alice_1,organization_alice
```

for request
```
alice, data1
```

Node-Casbin engine result is " false ".
Casbin-rs engine result is " true Reason: ["organization_alice","data1","allow"] ".

I hope to use a permission system with two dimensions: user role resource and user organization resource. I hope to obtain the final result of two dimensional permission systems (such as  true || false, true && true) through the definition of matchers. What should I do?

**Reply by z33gg** [reply]
@anhtuank7c

**Reply by hsluoyz** [reply]
@z33gg test with editor first: https://editor.casbin.org/

![image](https://github.com/user-attachments/assets/6bbb1654-1841-4b29-a14a-189610f7fe27)

**Comment by wuxiao-ling197** [comment]
how to use "custom functions" in matchers? how to write the policy demo_data? is anybody tried?

**Reply by hsluoyz** [reply]
@wuxiao-ling197 see docs "How to add a customized function": https://www.casbin.org/docs/function#how-to-add-a-customized-function

**Comment by tlyyxjz** [comment]
A recurring theme here is "my enforcement result is all false" / "I can't figure out the matcher / the `g()`". If it helps, the mental model is: every request `r` is tested against every policy `p`, and the `matcher` is a boolean test deciding whether that one pair matches. `g(r.sub, p.sub)` just means "is r.sub, directly or through roles, equal to p.sub". When you get all-false, it's almost always one of: matcher field names don't match the request/policy columns, a role link is missing from `g` (the role graph), or the effect/`p.eft` is off.

Out of curiosity (collecting real onboarding pain for a side project): would a small tool help that turns plain-language rules — "alice can read docs, bob owns project X, editors can write, nobody from team-ban is allowed" — into a valid, runnable `model.conf` + sample policy you can test, plus a plain-English explanation of why each request returns `true/false`? That's the exact confusion I keep seeing here, so I'm gauging whether it's worth building. What's been the single most confusing part for you? 🙂
